docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,185 @@
+
+* Status: accepted
+
+## Context and Problem Statement
+
+We can think of the policy matching algorithm as a function
+that maps a set of policies provided by the user to the following structure:
+
+```yaml
+Rules:
+  - Subset:
+      - Key: kuma.io/service
+        Not: false
+        Value: foo
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 15s
+      idleTimeout: 20s
+    Origin: [ ... ]
+  - Subset:
+      - Key: kuma.io/service
+        Not: false
+        Value: bar
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 20s
+      idleTimeout: 20s
+    Origin: [ ... ]
+```
+
+This structure is used by both "from" and "to" policies plugins.
+Policy plugins use the `Rules` list to compute `Conf` for the known destination.
+In example, if we want to know the `Conf` for outbound with `kuma.io/service: bar` tags,
+we're going to iterate over `Rules` and check if our outbound belongs to the `Rules[].Subset`.
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+type ResourceRule struct {
+    Resource core_model.ResourceMeta
+    Conf     interface{}
+    Origin   []core_model.ResourceMeta
+}
+```
+
+When targetRef has kind `MeshService` and uses `labels` we're going to produce multiple `ResourceRule` entities
+for each real `MeshService` that satisfies `labels`, i.e for the given `targetRef`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      labels:
+        k8s.kuma.io/namespace: finance
+    conf: conf1
+```
+
+we are going to fetch all `MeshService` resources, filter them by `k8s.kuma.io/namespace: finance` and produce N rules:
+
+```yaml
+resourceRules:
+  - resource:
+      name: finance-backend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+  - resource:
+      name: finance-frontend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+  - resource:
+      name: finance-db.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+```
+
+Policy plugins should implement a fallback logic when trying to get a conf for the given `MeshService{name: backend}`:
```

Comment:
How do we do this right now?

Reply:
I changed this part, it was never about fallback